[inductor] triton_utils.config_of: check for divisibility by 16, even when expr is not an Integer

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105743

TL;DR: triton_utils.config_of determines divisibility by 16 for each of the inputs to the kernel (pointer alignment for pointers, and divisibility by 16 for sizes). For sizes, the check previously could only return true if the expr representing the size was an integer. However, it's possible for non-integral exprs to be divisible by 16, e.g. for an expr like 16*s0.

Motivation: Knowledge about divisibility by 16 allows for vectorizing loads and stores, which can improve memory bandwidth. If we have, for example, kernels with shape [s0, 16] (dynamic batch size; static, divisible-by-16 other dimensions), we want to still be able to vectorize those loads and stores.

Dashboard results suggest that this improves dynamic shape training performance for timm, and possibly a small improvement for torchbench as well. More details are provided in a comment below.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

davidberard98 committed:
[inductor] triton_utils.config_of: check for divisibility by 16 even if expr is not an Integer

TL;DR: triton_utils.config_of determines divisibility by 16 for each of the inputs to the kernel (pointer alignment for pointers, and divisibility by 16 for sizes). For sizes, the check previously could only return true if the expr representing the size was an integer. However, it's possible for non-integral exprs to be divisible by 16, e.g. for an expr like 16*s0.

Motivation: Knowledge about divisibility by 16 allows for vectorizing loads and stores, which can improve memory bandwidth. If we have, for example, kernels with sizes [s0, 16] (dynamic batch size and static, divisible by 16 other dimensions), we want to still be able to vectorize those loads and stores.

[ghstack-poisoned]

davidberard98 committed:
Update on "[inductor] triton_utils.config_of: check for divisibility by 16 even if expr is not an Integer"


TL;DR: triton_utils.config_of determines divisibility by 16 for each of the inputs to the kernel (pointer alignment for pointers, and divisibility by 16 for sizes). For sizes, the check previously could only return true if the expr representing the size was an integer. However, it's possible for non-integral exprs to be divisible by 16, e.g. for an expr like 16*s0.

Motivation: Knowledge about divisibility by 16 allows for vectorizing loads and stores, which can improve memory bandwidth. If we have, for example, kernels with shape [s0, 16] (dynamic batch size; static, divisible-by-16 other dimensions), we want to still be able to vectorize those loads and stores.

cc voznesenskym penguinwu @EikanWang jgong5 @Guobing-Chen @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

davidberard98 commented:
lint and rebase on `322dff475c`, so I can compare perf run against the jul 21 benchmark run

davidberard98 commented:
Dashboard results: [link](https://hud.pytorch.org/benchmark/compilers?startTime=Sat%2C%2015%20Jul%202023%2003%3A22%3A54%20GMT&stopTime=Sat%2C%2022%20Jul%202023%2003%3A22%3A54%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/davidberard98/210/orig&lCommit=0766c61db2099d6c3de6ff638cf8e91fd4bed9d1&rBranch=main&rCommit=322dff475c0fa4a21288acc6de33b3fdac2621ea)

aakhundov commented on test/inductor/test_codegen_triton.py:
```diff
@@ -0,0 +1,77 @@
+
+    @inductor_config.patch("triton.divisible_by_16", True)
+    def test_config_of_sizearg(self):
+        two = sympy.Integer(4)
```
This should be `four`?

aakhundov commented on test/inductor/test_codegen_triton.py:
```diff
@@ -0,0 +1,77 @@
+            (2,),
+            triton_utils.config_of(
+                [
+                    SizeArg("A", two),
+                    SizeArg("B", eight),
+                    SizeArg("C", sixteen),
+                    SizeArg("D", s0),
+                    SizeArg("E", s1),
```
Perhaps add `# yes` / `# no` comments here, too?

aakhundov commented on test/inductor/test_codegen_triton.py:
```diff
@@ -0,0 +1,77 @@
+
+
+class TestCodegenTriton(TorchTestCase):
+    def setUp(self):
+        super().setUp()
+
+        class DummyModule(torch.nn.Module):
+            def forward(self, x):
+                return x * 2
+
+        self._gm = torch.fx.symbolic_trace(DummyModule())
+        self._graph = GraphLowering(self._gm)
+
+        self._stack = contextlib.ExitStack()
+        self._stack.enter_context(V.set_graph_handler(self._graph))
+
+    def tearDown(self):
+        self._stack.close()
+        super().tearDown()
```
Curious if we need this dummy module setup / teardown for this test.

davidberard98 replied:
I think we need it because V.graph is used inside triton_utils.config_of. A search in codegen/triton.py also shows V.graph 80 times, so I suspect if we added more unit tests for codegen many of those tests would also benefit from this setup/teardown.

aakhundov commented on torch/_inductor/codegen/triton_utils.py:
```diff
@@ -32,17 +30,12 @@ def is_aligned(x):
         if isinstance(x, TensorArg):
             return x.buffer not in V.graph.unaligned_buffers
         if isinstance(x, SizeArg):
-            if isinstance(x.expr, (int, sympy.Integer)):
```
Can it be that we're missing some important reason of having this check in place? cc @voznesenskym.

ezyang replied:
I think the old code is just wrong

davidberard98 committed:
Update on "[inductor] triton_utils.config_of: check for divisibility by 16, even when expr is not an Integer"


TL;DR: triton_utils.config_of determines divisibility by 16 for each of the inputs to the kernel (pointer alignment for pointers, and divisibility by 16 for sizes). For sizes, the check previously could only return true if the expr representing the size was an integer. However, it's possible for non-integral exprs to be divisible by 16, e.g. for an expr like 16*s0.

Motivation: Knowledge about divisibility by 16 allows for vectorizing loads and stores, which can improve memory bandwidth. If we have, for example, kernels with shape [s0, 16] (dynamic batch size; static, divisible-by-16 other dimensions), we want to still be able to vectorize those loads and stores.

Dashboard results suggest that this improves dynamic shape training performance for timm, and possibly a small improvement for torchbench as well. More details are provided in a comment below.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

davidberard98 commented:
@pytorchbot merge

davidberard98 commented:
@pytorchbot merge
